### PR TITLE
[Kernel] Add GPTQv2 format support for low-bit or asymmetric quantization, by adapting gptq_gemm

### PR DESCRIPTION
**Pull request by xxxxyu**
## Purpose

Add support for GPTQv2 format model checkpoints, by adapting `gptq_gemm` CUDA kernel with correct zero point handling. The condition is determined by `"checkpoint_format": "gptq_v2"` in `quantize_config.json`.

Currently, vllm treats GPTQv2 and GPTQv1 (the default GPTQ) format checkpoints equally, and causes gibberish output (repeated `!!!!`) with GPTQv2 format checkpoints (details in https://github.com/vllm-project/vllm/issues/26343).

<details>
<summary><h3>What is GPTQv2?</h3></summary>

GPTQv2 is both a **quantization algorithm** and a **checkpoint format**. This PR adds support for the **GPTQv2 checkpoint format**, not the quantization algorithm. Specifically:

- As a quantization algorithm: GPTQv2 (i.e., https://github.com/Intelligent-Computing-Lab-Panda/GPTAQ) provides potentially higher model accuracy than GPTQv1 (i.e., the default GPTQ). It is integrated to  the GPTQModel (https://github.com/ModelCloud/GPTQModel) library as an experimental feature (enabled by setting `v2=True`).
- As a checkpoint format: GPTQv2 adopts a slightly different checkpoint format with GPTQv1, in how to store the zero points of quantized weights. GPTQModel supports both GPTQv1 (by default) and GPTQv2 format (enabled by setting `format='gptq_v2'`). Checkpoints in GPTQv2 format will show `"checkpoint_format": "gptq_v2"` in `quantize_config.json`.

Also:

- A model quantized by the GPTQv2 algorithm could be saved in GPTQv1 format (details of v2 -> v1 conversion: 
https://github.com/ModelCloud/GPTQModel/blob/020ac04b74f6263f22491e6a6a034cb4fa5bf181/gptqmodel/utils/model.py#L625).
- A model quantized by other algorithms could be saved in GPTQv2 format (e.g., https://huggingface.co/BitDistiller/Qwen-8B-w2g64-gptq).

</details>

<details>
<summary><h3>Why support GPTQv2 format?</h3></summary>

GPTQv2, as a **checkpoint format**, provides **higher accuracy for asymmetrically (especially low-bit) quantized models**.

By default, GPTQModel uses GPTQv2 format internally, but converts to GPTQv1 format when storing the quantized checkpoint, mainly for compatibility purpose. However, [this conversion](https://github.com/ModelCloud/GPTQModel/blob/020ac04b74f6263f22491e6a6a034cb4fa5bf181/gptqmodel/utils/model.py#L625) is **not lossless**, and potentially harms model accuracy, especially for low-bit quantized models.

Specifically, the v2 -> v1 conversion requires subtracting 1 from the zero point (ranging [0, 2^b]), causing both `zero=0` and `zero=1` stored as the same value. For example, in INT2 quantization, this reduces the actual range of zero point values from {0,1,2,3} to {1,2,3} (both will be stored as {0,1,2} in GPTQv1 format).

</details>

<details>
<summary><h3>How to support GPTQv2 format?</h3></summary>

The only difference between GPTQv1 and GPTQv2 format, is how they stores the zero points. Specifically, GPTQv1 format subtracts 1 from zero points, and GPTQv2 format does not.

The current `gptq_gemm` kernel resumes the original zero points by adding 1 back to the zero points, when dequantizing weights. For example: https://github.com/vllm-project/vllm/blob/be22bb6f3dd7aaf8559a4a0a1beb98a37a5a8138/csrc/quantization/gptq/q_gemm.cu#L412

To support GPTQv2 format, in the newly added `gptq_gemm_v2` kernel, I simply removed all the `zero + 1` or `-zero -1` logic, and uses the zero points as is.

</details>

## Test Plan

I added/updated 2 tests to the codebase:

- I updated `tests/kernels/quantization/test_gptq.py`, by adding an opcheck of the newly added `gptq_gemm_v2` kernel.
- I added `tests/quantization/test_gptq_v2.py`, to test whether vllm correctly handles GPTQv2 format checkpoints and generates correct outputs.
	- Models: https://huggingface.co/BitDistiller/Qwen-8B-w2g64-gptq and https://huggingface.co/BitDistiller/Llama-3.1-8B-Instruct-w2g64-gptq.

For this PR specifically, I am testing with the following code:

```py
import torch
import logging
from transformers import AutoTokenizer
from vllm import LLM, SamplingParams

# Test models
MODELS = [
    "BitDistiller/Qwen-8B-w2g64-gptq",
    "BitDistiller/Llama-3.1-8B-Instruct-w2g64-gptq",
]

def test_gptq_v2_models():
    results = []
    
    for model_id in MODELS:
        # Collect output for this model
        model_output = []
        model_output.append(f"\nTesting {model_id}")
        model_output.append("-" * 50)
        
        try:
            llm = LLM(model=model_id, dtype=torch.float16, max_model_len=512)
            sampling_params = SamplingParams(max_tokens=128, temperature=0)
            
            # Prepare prompt
            tokenizer = AutoTokenizer.from_pretrained(model_id)
            messages = [
                {"role": "system", "content": "You are a helpful assistant."},
                {"role": "user", "content": "What is the meaning of life?"}
            ]
            
            # Handle thinking models
            enable_thinking = "Qwen" in model_id
            text = tokenizer.apply_chat_template(
                messages, tokenize=False, add_generation_prompt=True,
                enable_thinking=enable_thinking
            )
            
            model_output.append(f"Prompt:\n{text}")
            
            # Generate response
            output = llm.generate(text, sampling_params)
            response = output[0].outputs[0].text
            
            model_output.append(f"Response:\n{response}")
            
            del llm
            
        except Exception as e:
            model_output.append(f"Error: {str(e)}")
        
        results.append("\n".join(model_output))
    
    # Print all results at the end
    print("\n" + "=" * 60)
    print("RESULTS")
    print("=" * 60)
    for result in results:
        print(result)

if __name__ == "__main__":
    test_gptq_v2_models()
```

## Test Result

Running the above test code, I get:
image

The output now is correct.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: csrc/quantization/gptq/q_gemm_v2.cu [hunk 1]
```diff
@@ -0,0 +1,1564 @@
...
+https://github.com/qwopqwop200/GPTQ-for-LLaMa.
+
+This supports GPTQ v2 format checkpoints (checkpoint_format: 'gptq_v2'),
+by removing the v1-specific "zero + 1" logic during dequantization.
+Specifically, GPTQ v1 format checkpoints store (zero - 1), and need to + 1 at
+runtime during dequantization. GPTQ v2 format checkpoints store the zero point
+as is, and doesn't require + 1 at runtime. For more details, please refer to
+ModelCloud/GPTQModel:
+https://github.com/ModelCloud/GPTQModel/blob/020ac04b74f6263f22491e6a6a034cb4fa5bf181/gptqmodel/utils/model.py#L625
+*/
```
[the remark this answers is not in the document]
Comment (xxxxyu) [reply csrc/quantization/gptq/q_gemm_v2.cu]: I agree that this improves maintainability — if it is expected that this fallback GPTQ kernel will be updated (not likely since vllm already has several optimized GPTQ kernels like Marlin).

For now, maybe it is also OK to keep the duplicated code? I can continue to work on templating the functions if the reviewer think it's necessary.

**Comment by mgoin** [comment]
Could you have this format just use the gptq_marlin kernels? These are much more performant, should already support zero point, and are widely used by default for >sm80
Since pytorch only supports sm75 below sm80, I think we want to focus on marlin

**Comment by xxxxyu** [comment]
> Could you have this format just use the gptq_marlin kernels? These are much more performant, should already support zero point, and are widely used by default for >sm80 Since pytorch only supports sm75 below sm80, I think we want to focus on marlin

Thanks for your time : )

In my case, I am serving some low-bit (e.g., 2/3-bit) models. Since `gptq_marlin` only supports 4/8-bit models, vllm falls back to `gptq_gemm` and produces gibberish outputs.

Although 4/8-bit models are more popular, it is good to support 2/3-bit ones for efficiency consideration. On my machines, the 2-bit models runs faster than 4-bit ones. Some benchmark results (single A100 80GB):

```sh
# 2-bit GPTQv2
vllm bench latency --model BitDistiller/Qwen-8B-w2g64-gptq  --input-len 8192 --output-len 1 --batch-size 16 --dtype float16
Avg latency: 8.880313996163506 seconds

# 4-bit GPTQ
vllm bench latency --model JunHowie/Qwen3-8B-GPTQ-Int4 --input-len 8192 --output-len 1 --batch-size 16
Avg latency: 12.523354894698908 seconds

# 4-bit AWQ
vllm bench latency --model Qwen/Qwen3-8B-AWQ --input-len 8192 --output-len 1 --batch-size 16
Avg latency: 12.61037957224374 seconds
```

- About using marlin — I'm not quite sure about the feasibility to adapt `gptq_marlin` for 2/3-bit inference, but I'll take a look, and welcome for discussion.

- Plus, if it is not planned to support this after discussion, some doc update would be nice — the user might not tell the difference between v1/v2, and it took me quite a while to locate this issue.

**Comment by mgoin** [comment]
Ah I did not consider that gptq_gemm supports lower precisions than 4-bit. Okay @xxxxyu , I think this is a valid case if we can keep the code and binary size impact of this format addition down. 
First, I would like to prevent duplication and just add the format as a templated/conditional case in the kernel. We can use the same function and just add a new arg for the format, much like the exllama arg.
Second, I don't see any logic update to the gptq_marlin override to make it aware of this new format so we should make sure 4bit models in the new format aren't running with marlin incorrectly.
After those changes I think this PR should be in a reasonable state - thanks for the contribution!

File: tests/quantization/test_gptq_v2.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+from transformers import AutoTokenizer
+
+from vllm.model_executor.layers.quantization.gptq import GPTQLinearMethod
+
+# Asymmetrically quantized models, stored in GPTQ v1/v2 format
+# MODELS = [(model_id, checkpoint_format, enable_thinking)]
+MODELS = [
+    ("BitDistiller/Qwen-8B-w2g64-gptq", "gptq_v2", True),
+    ("BitDistiller/Llama-3.1-8B-Instruct-w2g64-gptq", "gptq_v2", False),
+]
```
Comment (mgoin) [inline tests/quantization/test_gptq_v2.py]: Can you use smaller models? I would appreciate 1B params at most for CI cost

Also do you need both? It seems that enable_thinking isn't related to quantization so not important to test here

**Review by mgoin — APPROVED** [review]
LGTM sorry for the delay @xxxxyu, just would like to prune the test time down a bit if possible

File: tests/quantization/test_gptq_v2.py (hunk 2 above)
Comment (xxxxyu) [reply tests/quantization/test_gptq_v2.py]: I can try quantizing a dummy model < 1B for the test. One is enough.

Comment (xxxxyu) [reply tests/quantization/test_gptq_v2.py]: Qwen3-1.7B can generate normal texts after simple 2-bit quantization, but Qwen3-0.6B cannot. So I uploaded https://huggingface.co/XXXXyu/Qwen3-1.7B-w2g64-gptq_v2 for testing. Model size is 1.0GB, should be OK?

Comment (xxxxyu) [reply tests/quantization/test_gptq_v2.py]: I can switch to 0.6B 2-bit too, but the checking logic would be simplified to only check if `!!!!` exists.

**Comment by xxxxyu** [comment]
Hi @mgoin, deprecating the buggy 4-bit `gptq_gemm` is causing some tests to fail. How do we deal with this?

**Comment by xxxxyu** [comment]
@mgoin I've replaced the 4-bit error message with a warning. Is this ready to merge?
